crashtracker/src/collector/crash_handler.rs

```diff
@@ -397,46 +394,36 @@ extern "C" fn handle_posix_sigaction(signum: i32, sig_info: *mut siginfo_t, ucon
+            SigHandler::SigDfl => {
+                // In the case of a default handler, we want to invoke it so that
+                // the core-dump can be generated.  Restoring the handler then
+                // re-raising the signal accomplishes that.
+                unsafe { signal::sigaction(*signal, sigaction) }
+                    .unwrap_or_else(|_| std::process::abort());
+                // Signals are only delivered once.
+                // In the case where we were invoked because of a crash, returning
+                // is technically UB but in practice re-invokes the crashing instr
+                // and re-raises the signal. In the case where we were invoked by
+                // `raise(SIGSEGV)` we need to re-raise the signal, or the default
+                // handler will never receive it.
+                unsafe { libc::raise(signum) };
+            }
+            SigHandler::SigIgn => (), // Return and ignore the signal.
+            SigHandler::Handler(f) => f(signum),
+            SigHandler::SigAction(f) => f(signum, sig_info, ucontext),
+        };
     } else {
-        unreachable!("The only signals we're registered for are SEGV and BUS")
-    };
-
-    // How we chain depends on what kind of handler we're chaining to.
-    // https://www.gnu.org/software/libc/manual/html_node/Signal-Handling.html
-    // https://man7.org/linux/man-pages/man2/sigaction.2.html
-    // Follow the approach here:
-    // https://stackoverflow.com/questions/6015498/executing-default-signal-handler
-    match old_sigaction.handler() {
-        SigHandler::SigDfl => {
-            // In the case of a default handler, we want to invoke it so that
-            // the core-dump can be generated.  Restoring the handler then
-            // re-raising the signal accomplishes that.
-            let signal = if signum == libc::SIGSEGV {
-                signal::SIGSEGV
-            } else if signum == libc::SIGBUS {
-                signal::SIGBUS
-            } else {
-                unreachable!("The only signals we're registered for are SEGV and BUS")
-            };
-            unsafe { signal::sigaction(signal, &old_sigaction) }
-                .unwrap_or_else(|_| std::process::abort());
-            // Signals are only delivered once.
-            // In the case where we were invoked because of a crash, returning
-            // is technically UB but in practice re-invokes the crashing instr
-            // and re-raises the signal. In the case where we were invoked by
-            // `raise(SIGSEGV)` we need to re-raise the signal, or the default
-            // handler will never receive it.
-            unsafe { libc::raise(signum) };
-        }
-        SigHandler::SigIgn => (), // Return and ignore the signal.
-        SigHandler::Handler(f) => f(signum),
-        SigHandler::SigAction(f) => f(signum, sig_info, ucontext),
-    };
+        eprintln!("Unexpected signal number {signum}, can't chain the handlers");
```

Comment:
What should we do here? abort?

Reply:
I guess this could happen if the user code changes the sigaction while we're in the middle of handling a signal?  I'm not sure how it might arise.